frontend/src/components/elements/PlotlyChart/CustomTheme.tsx

```diff
@@ -426,55 +197,178 @@ export function applyStreamlitThemeTemplateLayout(
     },
   }
 
-  // cant use merge. use assign because plotly already has properties defined.
-  assign(layout, streamlitTheme)
+  merge(layout, streamlitTheme)
 }
 
 /**
- * Applies the colorscale, colors for unique graphs, and discrete coloring
- * for in general through assigning properties in spec.data
- * @param data - spec.data
+ * Replace the colors that we are using from streamlit_plotly_theme.py.
+ * This is done so that we change colors based on the background color
+ * as the backend has no idea of the background color.
+ * @param spec the spec that we want to update
+ * @param theme
+ * @param elTheme element.theme
+ * @returns the updated spec with the correct theme colors
  */
-export function applyStreamlitThemeData(data: any, theme: Theme): void {
-  applyColorscale(data, theme)
-  applyUniqueGraphColorsData(data, theme)
-  applyDiscreteColors(data, theme)
-  data.forEach((entry: any) => {
-    if (entry.marker !== undefined) {
-      entry.marker.line = assign(entry.marker.line, {
-        width: 0,
-        color: theme.colors.transparent,
-      })
-    }
-  })
+function replaceCategoricalColors(
+  spec: string,
+  theme: Theme,
+  elTheme: string
+): string {
+  const CATEGORY_0 = "#000001"
+  const CATEGORY_1 = "#000002"
+  const CATEGORY_2 = "#000003"
+  const CATEGORY_3 = "#000004"
+  const CATEGORY_4 = "#000005"
+  const CATEGORY_5 = "#000006"
+  const CATEGORY_6 = "#000007"
+  const CATEGORY_7 = "#000008"
+  const CATEGORY_8 = "#000009"
+  const CATEGORY_9 = "#000010"
+
+  if (elTheme === "streamlit") {
+    const categoryColors = getCategoricalColorsArray(theme)
+    spec = spec.replaceAll(CATEGORY_0, categoryColors[0])
```

Comment:
nit: I'm a bit hesitant with all the repetitive `replaceAll` calls here, there might be a cleaner solution with some mapping & loop 🤔

Reply:
But thinking about this a bit more: the current implementation makes it very explicit so I'm fine with leaving it as is.